File: 24. 两两交换链表中的节点/go/main.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func main() {
	node := new(ListNode)
	node.Val = 1

	swapPairs(node)
}

func swapPairs(head *ListNode) *ListNode {
	if head == nil {
		return nil
	}

	pre := new(ListNode)
	pre.Next = head

	one := new(ListNode)
	two := new(ListNode)
	three := new(ListNode)
	four := new(ListNode)

	one = pre
	two = head
	three = head.Next
	if three != nil {
		four = three.Next
	}

	for two != nil {
		swapOnce(one, two, three, four)
		one = two
		two = four
		if four != nil && four.Next != nil {
			three = four.Next
		} else {
			three = nil
		}
		if three != nil && three.Next != nil {
			four = three.Next
		} else {
			four = nil
		}
	}

	return pre.Next
}

func swapOnce(pre *ListNode, left *ListNode, right *ListNode, next *ListNode) {
	if left != nil && right != nil {
		right.Next = left
		left.Next = next
		pre.Next = right
	}
}

File: 24. 两两交换链表中的节点/swift/main/main/main.swift
//
//  main.swift
//  main
//
//  Created by ByteDance on 4/25/25.
//

import Foundation

public class ListNode {
    public var val: Int
    public var next: ListNode?
    public init() { self.val = 0; self.next = nil; }
    public init(_ val: Int) { self.val = val; self.next = nil; }
    public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
}

class Solution {
    func swapPairs(_ head: ListNode?) -> ListNode? {
        let result = ListNode()
        result.next = head

        var p: ListNode? = result
        while p != nil && p?.next != nil && p?.next?.next != nil {
            let left = p?.next
            let right = p?.next?.next
            let nextP = right?.next

            p?.next = right
            right?.next = left
            left?.next = nextP

            p = left
        }

        return result.next
    }
}

let node1 = ListNode(1)
let node2 = ListNode(2)
let node3 = ListNode(3)
let node4 = ListNode(4)
node1.next = node2
node2.next = node3
node3.next = node4

let s = Solution()
s.swapPairs(node1)
